enarx-keep-sgx-shim/src/handler.rs

```diff
@@ -45,6 +45,35 @@ extern "C" {
 
 pub enum Context {}
 
+// Helper for allocation of untrusted memory in a Block.
+pub struct Cursor<'a>(Option<&'a mut [u8]>);
+
+impl<'a> Cursor<'a> {
+    pub fn new(buffer: &'a mut [u8]) -> Self {
+        Self(Some(buffer))
+    }
+
+    pub fn alloc<T>(&mut self, count: usize) -> Result<&'a mut [T], ()> {
+        let buf = self.0.take().unwrap();
+
+        let mid = {
+            let (padding, data, _) = unsafe { buf.align_to_mut::<T>() };
+
+            if data.len() < count {
+                return Err(());
+            }
+
+            padding.len() + size_of::<T>() * count
+        };
+
+        let (data, next) = buf.split_at_mut(mid);
+
+        self.0.replace(next);
+
+        Ok(unsafe { data.align_to_mut::<T>().1 })
+    }
+}
```

Comment:
I know I wrote this code. But it has some problems.

1. We should try to implement this without the `Option`. I forget why I added it. But it should be possible. This will remove the cases of `.unwrap()`.

2. `alloc()` either needs to take `T: Default` and use `Default::default()` to initialize all the elements of the returned array or it needs to be made `unsafe`. As it stands now, the function is safe but it returns an array to uninitialized elements. I think I'd go for the `unsafe` option.

3. The cursor should probably store a return value in case allocation fails. This means we can specify the `Err` return type as a type parameter and it will simplify error handling.

4. This probably belongs in the `sallyport` crate. If so, we can make the `buf` field private. The `Block` type would just grow a `fn cursor<'a, T>(err: T) -> Cursor<'a, T>` which would replace `Cursor::new()`.

Reply:
@npmccallum Are you sure about making the `buf` private? We can't use `if input.len() > self.block.buf.len()` in the `handler` if that field is private.

Reply:
I created #706 for this.